core/src/authorization/guards/action.guard.ts

```diff
@@ -0,0 +1,43 @@
+import type {CanActivate, ExecutionContext} from "@nestjs/common";
+import {Injectable} from "@nestjs/common";
+import {Reflector} from "@nestjs/core";
+import {GqlExecutionContext} from "@nestjs/graphql";
+
+import {MemberRepository} from "$repositories";
+
+import {JwtAuthPayloadSchema} from "../../authentication/types";
+import {AuthorizationService} from "../authorization.service";
+import {MetadataKeys} from "../authorization.types";
+
+@Injectable()
+export class ActionGuard implements CanActivate {
+    constructor(
+        private readonly reflector: Reflector,
+        private readonly authorizationService: AuthorizationService,
+        private readonly memberRepository: MemberRepository,
+    ) {}
+
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+        const requiredActions = this.reflector.getAllAndOverride<string[] | undefined>(MetadataKeys.Actions, [
+            context.getHandler(),
+            context.getClass(),
+        ]);
+
+        if (!requiredActions?.length) return true;
+
+        const ctx = GqlExecutionContext.create(context).getContext();
+        if (!ctx?.req.user) return false;
+
+        const data = JwtAuthPayloadSchema.safeParse(ctx.req.user);
+        if (!data.success || !data.data.currentOrganizationId) return false;
+
+        const member = await this.memberRepository.getOrNull({
+            where: {userId: data.data.userId, organizationId: data.data.currentOrganizationId},
+        });
+        if (!member) return false;
+
+        const memberActions = await this.authorizationService.getMemberActions(member.id);
+        if (!requiredActions.some(requiredAction => memberActions.includes(requiredAction))) return false;
```

Comment:
For this decorator, are `requiredActions` all required, or just one of them is required?

Currently, this `.some` just checks for at least one

Reply:
Just one of them. Acts as an OR. Not sure if that’s what we want but that’s what I assumed.

Reply:
That's fine, we should just update the guard interface to be more explicit about that

Maybe something like
```ts
@Action({ oneOf: ["action1", "action2"] })
someFunction(): void {

}
```

Reply:
Then we could also add `allOf`

```ts
@Action({ allOf: ["action1", "action2"] })
someFunction(): void {

}
```